python/pyiceberg/expressions/base.py

```diff
@@ -342,151 +313,197 @@ def __str__(self) -> str:
-        return BoundNotIn(self.term, *self.literals)
+class BoundIn(BoundSetPredicate[T]):
+    def __new__(cls, term: BoundTerm[T], literals: set[Literal[T]]) -> BooleanExpression:
+        count = len(literals)
+        if count == 0:
+            return AlwaysFalse()
+        if count == 1:
+            return BoundEq(term, literals.pop())
+        else:
+            return super().__new__(cls)
 
+    def __invert__(self) -> BooleanExpression:
+        return BoundNotIn(self.term, self.literals)
 
-class In(UnboundPredicate[T]):
-    def _validate_literals(self):  # pylint: disable=W0238
-        if not self.literals:
-            raise AttributeError("In must contain at least 1 literal.")
 
-    def __invert__(self) -> NotIn[T]:
-        return NotIn(self.term, *self.literals)
+class BoundNotIn(BoundSetPredicate[T]):
+    def __new__(cls, term: BoundTerm[T], literals: set[Literal[T]]) -> BooleanExpression:
+        count = len(literals)
+        if count == 0:
+            return AlwaysTrue()
+        if count == 1:
+            return BoundNotEq(term, literals.pop())
+        else:
+            return super().__new__(cls)
 
-    def bind(self, schema: Schema, case_sensitive: bool) -> BoundIn[T]:
-        bound_ref = self.term.bind(schema, case_sensitive)
-        return BoundIn(bound_ref, *tuple(lit.to(bound_ref.field.field_type) for lit in self.literals))  # type: ignore
+    def __invert__(self) -> BooleanExpression:
+        return BoundIn(self.term, self.literals)
 
 
-class BoundNotIn(BoundPredicate[T]):
-    def _validate_literals(self):  # pylint: disable=W0238
-        if not self.literals:
-            raise AttributeError("BoundNotIn must contain at least 1 literal.")
+class In(SetPredicate[T]):
+    as_bound = BoundIn
 
-    def __invert__(self) -> BoundIn[T]:
-        return BoundIn(self.term, *self.literals)
+    def __new__(cls, term: UnboundTerm[T], literals: tuple[Literal[T], ...]) -> BooleanExpression:
+        count = len(literals)
+        if count == 0:
+            return AlwaysFalse()
+        if count == 1:
+            return Eq(term, literals[0])
+        else:
+            return super().__new__(cls)
 
+    def __invert__(self) -> BooleanExpression:
+        return NotIn(self.term, self.literals)
 
-class NotIn(UnboundPredicate[T]):
-    def _validate_literals(self):  # pylint: disable=W0238
-        if not self.literals:
-            raise AttributeError("NotIn must contain at least 1 literal.")
 
-    def __invert__(self) -> In[T]:
-        return In(self.term, *self.literals)
+class NotIn(SetPredicate[T]):
+    as_bound = BoundNotIn
 
-    def bind(self, schema: Schema, case_sensitive: bool) -> BoundNotIn[T]:
-        bound_ref = self.term.bind(schema, case_sensitive)
-        return BoundNotIn(bound_ref, *tuple(lit.to(bound_ref.field.field_type) for lit in self.literals))  # type: ignore
+    def __new__(cls, term: UnboundTerm[T], literals: tuple[Literal[T], ...]) -> BooleanExpression:
```

Comment:
```suggestion
    def __new__(cls, term: UnboundTerm[T], literals: *Literal[T]) -> BooleanExpression:
```
See the comment at `class In(SetPredicate[T]):`